File: gui-lov/src/lib/fileStore.ts
/**
 * Storages, and the files in them.
 *
 * A campaign's `recipients.files` takes URLs or minio bucket/keys, and an author should not have
 * to know which — they upload a file and copy what comes back. Uploading happens at that moment,
 * so the reference is real and stable immediately, which is why nothing else in the app needs to
 * change: it is text from then on. Paste it into a body, or name it as an environment variable.
 *
 * Storages are **named and project-scoped**, not per-environment. Uploading happens while
 * authoring, so which environment is selected has no bearing on the result — and having two at
 * once is the point, because migrating from minio to a file-service pod means running both for a
 * while.
 */

/** Extensions this platform accepts for a recipients file. */
export const KNOWN_EXTENSIONS = ["csv", "txt", "xls", "xlsx"] as const;

export type StoreKind = "s3" | "http";
export type ReferenceForm = "key" | "url";

/** A storage as the server describes it. Never carries a secret — there is no field for one. */
export interface StoreDef {
  id: string;
  project_id: string;
  name: string;
  kind: StoreKind;
  endpoint: string;
  bucket?: string;
  prefix: string;
  access_key?: string;
  reference: ReferenceForm;
  region: string;
  /** A secret is stored — so the form shows "••••••• (set)" and only overwrites it if you type. */
  has_secret: boolean;
  has_token: boolean;
  /** `env` storages are read-only: the shell that exported them is the only place to change them. */
  source: "project" | "env";
  created_at: string;
  updated_at: string;
}

/** What the form sends. The two secrets are omitted rather than blanked when unchanged. */
export interface StoreDraft {
  name: string;
  kind: StoreKind;
  endpoint: string;
  bucket?: string;
  prefix?: string;
  access_key?: string;
  secret_key?: string;
  token?: string;
  reference?: ReferenceForm;
  region?: string;
}

export interface CheckStep {
  step: string;
  /** `null` means an earlier step failed and this one never ran — not the same as failing. */
  ok: boolean | null;
  detail?: string;
}

export interface CheckResult {
  ok: boolean;
  steps: CheckStep[];
}

export interface StoredFile {
  name: string;
  key: string;
  /** What the author pastes into a test: a URL, or `bucket/key`. */
  reference: string;
  size: number;
  uploaded_at?: string;
}

// ---------------------------------------------------------------- storages

export function isReadOnly(store: StoreDef | undefined): boolean {
  return store?.source === "env";
}

/** "minio · sat-fixtures @ 127.0.0.1:9000" — never a credential. */
export function storeLabel(store: StoreDef): string {
  const host = store.endpoint.replace(/^https?:\/\//, "").replace(/\/$/, "");
  if (store.kind === "s3") {
    return store.bucket ? `minio · ${store.bucket} @ ${host}` : `minio · ${host}`;
  }
  return `file service · ${host}`;
}

/** A blank storage of this kind, with the reference form that kind can actually deliver. */
export function blankDraft(kind: StoreKind): StoreDraft {
  return {
    name: "",
    kind,
    endpoint: "",
    bucket: kind === "s3" ? "" : undefined,
    prefix: "",
    access_key: kind === "s3" ? "" : undefined,
    secret_key: "",
    token: kind === "http" ? "" : undefined,
    reference: defaultReference(kind),
  };
}

/**
 * minio hands over a key, a file service its own URL.
 *
 * A URL pointing at an in-cluster minio is an RFC1918 address, which is exactly what this
 * platform's SSRF guard rejects — so defaulting an S3 storage to a URL would default it to
 * broken.
 */
export function defaultReference(kind: StoreKind): ReferenceForm {
  return kind === "http" ? "url" : "key";
}

/**
 * A draft for editing an existing storage.
 *
 * **The secrets come back empty, meaning "leave them alone".** The server never sends them, so
 * the form cannot echo them; sending an empty string would clear a working credential on every
 * rename.
 */
export function draftFromStore(store: StoreDef): StoreDraft {
  return {
    name: store.name,
    kind: store.kind,
    endpoint: store.endpoint,
    bucket: store.bucket ?? "",
    prefix: store.prefix,
    access_key: store.access_key ?? "",
    secret_key: undefined,
    token: undefined,
    reference: store.reference,
    region: store.region,
  };
}

/**
 * What still has to be filled in before this can be saved or tested.
 *
 * Author-facing words, not field names — this is shown next to a disabled button, so it has to
 * say what to do rather than name an identifier.
 */
export function missingFields(draft: StoreDraft, existing?: StoreDef): string[] {
  const missing: string[] = [];
  if (!draft.name.trim()) missing.push("a name");
  if (!draft.endpoint.trim()) missing.push("an endpoint");
  if (draft.kind === "s3") {
    if (!draft.bucket?.trim()) missing.push("a bucket");
    if (!draft.access_key?.trim()) missing.push("an access key");
    // Already stored counts as present: the form was never given it, and demanding it again
    // would make every edit require re-typing a credential.
    if (!draft.secret_key?.trim() && !existing?.has_secret) missing.push("a secret key");
  }
  return missing;
}

/** The endpoint mistake worth catching before a round-trip: no scheme. */
export function endpointHint(endpoint: string): string | undefined {
  const e = endpoint.trim();
  if (!e || e.startsWith("http://") || e.startsWith("https://")) return undefined;
  return `Needs a scheme — try http://${e}`;
}

/**
 * What to send: the secrets are dropped when untouched, so absence means "keep the stored one".
 *
 * An empty string is deliberately *not* dropped when the author cleared a field they had typed
 * into, because clearing is a real request. Only `undefined` means "unchanged".
 */
export function toPayload(draft: StoreDraft): StoreDraft {
  const out: StoreDraft = { ...draft, name: draft.name.trim(), endpoint: draft.endpoint.trim() };
  if (out.secret_key === undefined || out.secret_key === "") delete out.secret_key;
  if (out.token === undefined || out.token === "") delete out.token;
  if (out.kind === "http") {
    delete out.bucket;
    delete out.access_key;
  }
  return out;
}

// ---------------------------------------------------------------- files

export function extensionOf(name: string): string {
  const i = name.lastIndexOf(".");
  return i > 0 ? name.slice(i + 1).toLowerCase() : "";
}

export function extensionIsKnown(name: string): boolean {
  return (KNOWN_EXTENSIONS as readonly string[]).includes(extensionOf(name));
}

/**
 * Advice, never a refusal.
 *
 * "wrong extension → 400" is a case an author wants to write, so blocking the upload would make
 * that test unwritable. The store holds whatever it is given.
 */
export function extensionWarning(name: string): string | undefined {
  if (extensionIsKnown(name)) return undefined;
  const ext = extensionOf(name);
  return `${ext ? `.${ext}` : "No extension"} — this API accepts ${KNOWN_EXTENSIONS.map(
    (e) => `.${e}`,
  ).join(" ")}, so it will reject this. Fine if that is the case you are writing.`;
}

export function formatSize(bytes: number): string {
  if (!Number.isFinite(bytes) || bytes < 0) return "—";
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

/**
 * What to call the thing you copy.
 *
 * "location" rather than "bucket / key": a key is genuinely not a URL, so calling it one would
 * be a small lie an author reads every time — but *bucket* and *key* are words from one
 * particular storage product, and they belong on the page that explains them rather than in a
 * toast. The precise form is shown there.
 */
export function referenceLabel(reference: ReferenceForm | undefined): string {
  if (reference === "key") return "location";
  if (reference === "url") return "URL";
  return "link";
}

/** Newest first, then by name — a list you scan for what you just uploaded. */
export function sortFiles(files: StoredFile[]): StoredFile[] {
  return [...files].sort((a, b) => {
    const at = a.uploaded_at ?? "";
    const bt = b.uploaded_at ?? "";
    if (at !== bt) return at < bt ? 1 : -1;
    return a.name.localeCompare(b.name);
  });
}

export function matchesQuery(file: StoredFile, query: string): boolean {
  const q = query.trim().toLowerCase();
  if (!q) return true;
  // The reference too: pasting a URL back in to find which file it was is a real move.
  return file.name.toLowerCase().includes(q) || file.reference.toLowerCase().includes(q);
}

File: api/src/api/file_store.rs
//! Configuring a storage, and using it.
//!
//! Satyanaash proxies every call to a store so its credentials stay in this process. A saved
//! store keeps them in the database, write-only — see `db/repositories/file_stores.rs`. A store
//! defined by environment variables is listed too, read-only, because the shell that exported
//! it is the only place it can be changed.
//!
//! Stores are **named and project-scoped**, not per-environment: uploading happens while
//! authoring and yields a literal, so which environment is selected has no bearing on the
//! result — and having two at once is the point, since migrating from minio to a file service
//! means running both for a while.

use std::sync::Arc;

use axum::{
    extract::{Multipart, Path, Query, State},
    http::StatusCode,
    Json,
};
use serde::{Deserialize, Serialize};

use crate::db::models::{CreateFileStore, FileStore, UpdateFileStore};
use crate::db::repositories::FileStoreRepository;
use crate::error::AppError;
use crate::files::{
    config, open, sanitise, FileStore as Store, Reference, StoreConfig, StoreKind, StoredFile,
    MAX_UPLOAD_BYTES,
};

#[derive(Clone)]
pub struct FileStoreState {
    pub repo: Arc<dyn FileStoreRepository>,
}

/// GET /api/v1/projects/{id}/file-stores
///
/// Saved stores first, then any the environment defines. The env ones come last and are marked,
/// because a read-only entry mixed in among editable ones would offer a Save that silently did
/// nothing.
pub async fn list_stores(
    State(state): State<FileStoreState>,
    Path(project_id): Path<String>,
) -> Result<Json<Vec<FileStore>>, AppError> {
    let mut stores = state.repo.list_by_project(&project_id).await?;
    stores.extend(
        config::from_environment()
            .into_iter()
            .map(|(name, cfg)| env_store(&project_id, &name, &cfg)),
    );
    Ok(Json(stores))
}

fn kind_str(kind: StoreKind) -> &'static str {
    match kind {
        StoreKind::S3 => "s3",
        StoreKind::Http => "http",
    }
}

fn reference_str(reference: Reference) -> &'static str {
    match reference {
        Reference::Key => "key",
        Reference::Url => "url",
    }
}

/// An environment-defined store, in the shape the list returns.
///
/// The id is prefixed `env:` so every route can tell at a glance that there is no database row
/// behind it, and so a saved store's UUID can never collide with one. The name is the variable
/// prefix itself, which says where it came from without a second naming scheme to invert.
fn env_store(project_id: &str, name: &str, cfg: &StoreConfig) -> FileStore {
    FileStore {
        id: format!("env:{name}"),
        project_id: project_id.to_string(),
        name: name.to_string(),
        kind: kind_str(cfg.kind).to_string(),
        endpoint: cfg.endpoint.clone(),
        bucket: cfg.bucket.clone(),
        prefix: cfg.prefix.clone(),
        access_key: cfg.access_key.clone(),
        reference: reference_str(cfg.reference).to_string(),
        region: cfg.region.clone(),
        has_secret: cfg.secret_key.is_some(),
        has_token: cfg.token.is_some(),
        source: "env".to_string(),
        // No row, so no history. Now rather than a fabricated past date, which would sort
        // convincingly wrongly if the list is ever ordered by age.
        created_at: chrono::Utc::now(),
        updated_at: chrono::Utc::now(),
    }
}

pub async fn create_store(
    State(state): State<FileStoreState>,
    Path(project_id): Path<String>,
    Json(input): Json<CreateFileStore>,
) -> Result<(StatusCode, Json<FileStore>), AppError> {
    Ok((StatusCode::CREATED, Json(state.repo.create(&project_id, input).await?)))
}

pub async fn update_store(
    State(state): State<FileStoreState>,
    Path(id): Path<String>,
    Json(input): Json<UpdateFileStore>,
) -> Result<Json<FileStore>, AppError> {
    refuse_env(&id, "changed")?;
    Ok(Json(state.repo.update(&id, input).await?))
}

pub async fn delete_store(
    State(state): State<FileStoreState>,
    Path(id): Path<String>,
) -> Result<StatusCode, AppError> {
    refuse_env(&id, "deleted")?;
    state.repo.delete(&id).await?;
    Ok(StatusCode::NO_CONTENT)
}

/// A store from the environment cannot be edited here, and says where it can be.
fn refuse_env(id: &str, verb: &str) -> Result<(), AppError> {
    if let Some(name) = id.strip_prefix("env:") {
        return Err(AppError::BadRequest(format!(
            "\"{name}\" comes from environment variables and cannot be {verb} here — edit \
             {name}_* where satyanaash is started, and restart it"
        )));
    }
    Ok(())
}

/// The config behind a store id, whether saved or from the environment.
async fn config_for(state: &FileStoreState, id: &str) -> Result<StoreConfig, AppError> {
    if let Some(name) = id.strip_prefix("env:") {
        return config::from_environment()
            .into_iter()
            .find(|(n, _)| n == name)
            .map(|(_, cfg)| cfg)
            .ok_or_else(|| {
                AppError::NotFound(format!(
                    "no storage called \"{name}\" in the environment any more — it may have been \
                     removed since this page loaded"
                ))
            });
    }
    state
        .repo
        .config_for(id)
        .await?
        .ok_or_else(|| AppError::NotFound(format!("no storage with id {id}")))
}

// ------------------------------------------------------------- testing a storage

/// One step of a storage check.
#[derive(Debug, Serialize)]
pub struct CheckStep {
    pub step: String,
    /// `None` when an earlier step failed and this one was never attempted — which is not the
    /// same as failing, and a cross would send the author to the wrong place.
    pub ok: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct CheckResult {
    pub ok: bool,
    pub steps: Vec<CheckStep>,
}

/// The probe object a check writes and removes.
///
/// Named so that one left behind in a bucket explains itself.
const PROBE: &str = "satyanaash-storage-check.txt";

/// A real round-trip, not a ping.
///
/// Listing proves the endpoint, the credentials and the bucket. It does **not** prove writing,
/// which is a separate permission and the one that matters — so the check uploads a small
/// object and removes it again, reporting the three steps separately. One tick over the lot
/// would hide a read-only key until the first real upload.
async fn run_check(cfg: &StoreConfig) -> CheckResult {
    let mut steps = Vec::new();

    let store = match open(cfg) {
        Ok(s) => s,
        Err(e) => {
            steps.push(CheckStep {
                step: "Open the storage".into(),
                ok: Some(false),
                detail: Some(plain(&e)),
            });
            return CheckResult { ok: false, steps };
        }
    };

    let reach = format!("Reach {} and list it", cfg.endpoint);
    let listed = store.list().await;
    match &listed {
        Ok(files) => steps.push(CheckStep {
            step: reach,
            ok: Some(true),
            detail: Some(match files.len() {
                0 => "empty".to_string(),
                1 => "1 file".to_string(),
                n => format!("{n} files"),
            }),
        }),
        Err(e) => steps.push(CheckStep { step: reach, ok: Some(false), detail: Some(plain(e)) }),
    }
    if listed.is_err() {
        // Not attempted, and said so rather than shown as failures for steps that never ran.
        steps.push(CheckStep { step: "Upload a probe file".into(), ok: None, detail: None });
        steps.push(CheckStep { step: "Delete it again".into(), ok: None, detail: None });
        return CheckResult { ok: false, steps };
    }

    let put = store.put(PROBE, b"satyanaash storage check\n".to_vec(), Some("text/plain")).await;
    match &put {
        Ok(f) => steps.push(CheckStep {
            step: "Upload a probe file".into(),
            ok: Some(true),
            // What an author would paste, shown once here so the Reference choice is visible
            // before any test depends on it.
            detail: Some(f.reference.clone()),
        }),
        Err(e) => steps.push(CheckStep {
            step: "Upload a probe file".into(),
            ok: Some(false),
            detail: Some(plain(e)),
        }),
    }

    match put {
        Ok(f) => match store.delete(&f.key).await {
            Ok(()) => {
                steps.push(CheckStep { step: "Delete it again".into(), ok: Some(true), detail: None })
            }
            Err(e) => steps.push(CheckStep {
                step: "Delete it again".into(),
                // Called out rather than shrugged off: a probe left behind is litter, and a
                // store that cannot delete cannot be tidied from the Files screen either.
                ok: Some(false),
                detail: Some(format!("{} — {PROBE} is still there", plain(&e))),
            }),
        },
        Err(_) => steps.push(CheckStep { step: "Delete it again".into(), ok: None, detail: None }),
    }

    let ok = steps.iter().all(|s| s.ok == Some(true));
    CheckResult { ok, steps }
}

/// POST /api/v1/file-stores/{id}/test
pub async fn test_store(
    State(state): State<FileStoreState>,
    Path(id): Path<String>,
) -> Result<Json<CheckResult>, AppError> {
    let cfg = config_for(&state, &id).await?;
    Ok(Json(run_check(&cfg).await))
}

/// POST /api/v1/projects/{id}/file-stores/test — check a configuration that is not saved yet.
///
/// So an author finds out whether it works *before* committing it, which is the difference
/// between a form that helps and one that records a guess.
pub async fn test_draft(
    Path(_project_id): Path<String>,
    Json(input): Json<CreateFileStore>,
) -> Result<Json<CheckResult>, AppError> {
    Ok(Json(run_check(&draft_config(input)?).await))
}

fn draft_config(input: CreateFileStore) -> Result<StoreConfig, AppError> {
    let endpoint = input.endpoint.trim().trim_end_matches('/').to_string();
    if endpoint.is_empty() {
        return Err(AppError::BadRequest("a storage needs an endpoint".to_string()));
    }
    let kind = match input.kind.as_str() {
        "http" => StoreKind::Http,
        _ => StoreKind::S3,
    };
    Ok(StoreConfig {
        kind,
        endpoint,
        bucket: input.bucket,
        prefix: match input.prefix.as_deref().map(str::trim).filter(|p| !p.is_empty()) {
            Some(p) => format!("{}/", p.trim_matches('/')),
            None => String::new(),
        },
        access_key: input.access_key,
        secret_key: input.secret_key,
        token: input.token,
        reference: match input.reference.as_deref() {
            Some("url") => Reference::Url,
            Some("key") => Reference::Key,
            // minio hands over a key: a URL to an in-cluster minio is an RFC1918 address,
            // which is exactly what the platform's SSRF guard rejects.
            _ => match kind {
                StoreKind::Http => Reference::Url,
                StoreKind::S3 => Reference::Key,
            },
        },
        region: input.region.unwrap_or_else(|| "us-east-1".to_string()),
    })
}

/// POST /api/v1/projects/{id}/file-stores/buckets — what this credential can see.
///
/// Takes a draft rather than a saved id, because the whole point is to answer it *while* someone
/// is filling the form in and does not yet know what to type.
pub async fn list_buckets(
    Path(_project_id): Path<String>,
    Json(input): Json<CreateFileStore>,
) -> Result<Json<Vec<String>>, AppError> {
    Ok(Json(open(&draft_config(input)?)?.buckets().await?))
}

#[derive(Debug, Deserialize)]
pub struct CreateBucketRequest {
    #[serde(flatten)]
    pub store: CreateFileStore,
    /// The bucket to make. Separate from the draft's own `bucket` so this cannot quietly create
    /// something other than what the author asked for.
    pub bucket: String,
}

/// POST /api/v1/projects/{id}/file-stores/create-bucket
pub async fn create_bucket(
    Path(_project_id): Path<String>,
    Json(input): Json<CreateBucketRequest>,
) -> Result<StatusCode, AppError> {
    let name = input.bucket.trim().to_string();
    if name.is_empty() {
        return Err(AppError::BadRequest("which bucket? a name is required".to_string()));
    }
    open(&draft_config(input.store)?)?.create_bucket(&name).await?;
    Ok(StatusCode::CREATED)
}

// ------------------------------------------------------------- files in a storage

#[derive(Debug, Deserialize)]
pub struct KeyQuery {
    pub key: String,
}

async fn store_for(state: &FileStoreState, id: &str) -> Result<Box<dyn Store>, AppError> {
    open(&config_for(state, id).await?)
}

/// GET /api/v1/file-stores/{id}/files
pub async fn list_files(
    State(state): State<FileStoreState>,
    Path(id): Path<String>,
) -> Result<Json<Vec<StoredFile>>, AppError> {
    Ok(Json(store_for(&state, &id).await?.list().await?))
}

/// POST /api/v1/file-stores/{id}/files — multipart, one or more `file` parts.
///
/// Several in one request, because "select one or more files" is a single action and five
/// separate failures for it would be five things to make sense of.
pub async fn upload(
    State(state): State<FileStoreState>,
    Path(id): Path<String>,
    mut form: Multipart,
) -> Result<(StatusCode, Json<Vec<StoredFile>>), AppError> {
    let store = store_for(&state, &id).await?;
    let mut stored = Vec::new();

    while let Some(field) = form
        .next_field()
        .await
        .map_err(|e| AppError::BadRequest(format!("could not read the upload: {e}")))?
    {
        let name = field.file_name().map(sanitise).unwrap_or_default();
        if name.is_empty() {
            continue;
        }
        let mime = field.content_type().map(str::to_string);
        let bytes = field
            .bytes()
            .await
            .map_err(|e| AppError::BadRequest(format!("could not read \"{name}\": {e}")))?;

        if bytes.len() as u64 > MAX_UPLOAD_BYTES {
            return Err(AppError::BadRequest(format!(
                "\"{}\" is {:.1} MB — this server accepts up to {} MB, because an upload passes \
                 through its memory",
                name,
                bytes.len() as f64 / 1_048_576.0,
                MAX_UPLOAD_BYTES / 1_048_576
            )));
        }
        if bytes.is_empty() {
            return Err(AppError::BadRequest(format!("\"{name}\" is empty")));
        }

        // The extension is deliberately not checked: `wrong extension → 400` is a case an
        // author wants to write, so the store must hold whatever it is given. The UI flags it.
        stored.push(store.put(&name, bytes.to_vec(), mime.as_deref()).await?);
    }

    if stored.is_empty() {
        return Err(AppError::BadRequest(
            "no files in the upload — expected one or more `file` parts".to_string(),
        ));
    }
    Ok((StatusCode::CREATED, Json(stored)))
}

/// DELETE /api/v1/file-stores/{id}/files?key=…
///
/// The key is a query parameter rather than a path segment because it contains slashes, and a
/// wildcard route would have to guess where it begins.
pub async fn delete_file(
    State(state): State<FileStoreState>,
    Path(id): Path<String>,
    Query(q): Query<KeyQuery>,
) -> Result<StatusCode, AppError> {
    if q.key.trim().is_empty() {
        return Err(AppError::BadRequest("which file? `key` is required".to_string()));
    }
    store_for(&state, &id).await?.delete(&q.key).await?;
    Ok(StatusCode::NO_CONTENT)
}

fn plain(e: &AppError) -> String {
    format!("{e}")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn an_environment_store_cannot_be_edited_and_says_where_it_can_be() {
        // Offering a Save that silently did nothing would be worse than refusing.
        let e = refuse_env("env:SAT_FILESTORE_DEV", "changed").unwrap_err();
        let msg = format!("{e}");
        assert!(msg.contains("SAT_FILESTORE_DEV_*"), "{msg}");
        assert!(msg.contains("restart"), "{msg}");
        // A saved store is untouched by the guard.
        assert!(refuse_env("2f9c-uuid", "changed").is_ok());
    }

    #[test]
    fn a_draft_takes_its_reference_form_from_its_kind() {
        let s3 = draft_config(CreateFileStore {
            name: "m".into(),
            kind: "s3".into(),
            endpoint: "http://h:9000".into(),
            bucket: Some("b".into()),
            prefix: Some("/fx/".into()),
            access_key: None,
            secret_key: None,
            token: None,
            reference: None,
            region: None,
        })
        .unwrap();
        // A URL to an in-cluster minio is RFC1918, which is the case the SSRF guard rejects.
        assert_eq!(s3.reference, Reference::Key);
        assert_eq!(s3.prefix, "fx/");

        let http = draft_config(CreateFileStore {
            name: "f".into(),
            kind: "http".into(),
            endpoint: "https://files/".into(),
            bucket: None,
            prefix: None,
            access_key: None,
            secret_key: None,
            token: None,
            reference: None,
            region: None,
        })
        .unwrap();
        assert_eq!(http.reference, Reference::Url);
        assert_eq!(http.endpoint, "https://files");
    }

    #[tokio::test]
    async fn a_check_reports_the_step_that_failed_and_does_not_pretend_the_rest_ran() {
        // A dead port, so listing cannot succeed.
        let cfg = draft_config(CreateFileStore {
            name: "f".into(),
            kind: "http".into(),
            endpoint: "http://127.0.0.1:1".into(),
            bucket: None,
            prefix: None,
            access_key: None,
            secret_key: None,
            token: None,
            reference: None,
            region: None,
        })
        .unwrap();

        let result = run_check(&cfg).await;
        assert!(!result.ok);
        assert_eq!(result.steps.len(), 3);
        assert_eq!(result.steps[0].ok, Some(false));
        assert!(result.steps[0].detail.as_deref().unwrap().contains("127.0.0.1:1"));
        // `None`, not `false`: these never ran, and a cross would send the author looking at
        // permissions when the endpoint is what is wrong.
        assert_eq!(result.steps[1].ok, None);
        assert_eq!(result.steps[2].ok, None);
    }

    #[test]
    fn an_environment_store_is_listed_read_only_with_no_secret_in_it() {
        let cfg = StoreConfig {
            kind: StoreKind::S3,
            endpoint: "http://127.0.0.1:9000".into(),
            bucket: Some("sat-fixtures".into()),
            prefix: "fixtures/".into(),
            access_key: Some("AK".into()),
            secret_key: Some("supersecret".into()),
            token: None,
            reference: Reference::Key,
            region: "us-east-1".into(),
        };
        let store = env_store("p1", "SAT_FILESTORE_DEV", &cfg);
        assert_eq!(store.id, "env:SAT_FILESTORE_DEV");
        assert_eq!(store.source, "env");
        assert!(store.has_secret);
        // Same guarantee as a saved store: the secret has no field to travel in.
        let json = serde_json::to_string(&store).unwrap();
        assert!(!json.contains("supersecret"), "{json}");
    }
}
